### PR TITLE
Block usage of enum types but allow them as relaxed types

### PR DESCRIPTION
**Pull request by daniel-chambers**
We've never supported enum types or unions of literals, but now we explicitly identify and block them with good error messages. Previously most instances of enums and unions of literals would have been blocked by other code (most enums have index signatures, and unions of literals are unions), but the error messages were unfriendly.

Now by explicitly identifying these kinds of types, we can display proper error messages mentioning enum types and also allow them under relaxed types. This fixes issues where unions of literals that previously would have _not_ been allowed under relaxed types now are allowed (for example, `"one" | "two" | undefined` was not allowed as a relaxed type).
This also lays the groundwork for properly supporting these types as enum types in the future.

The code around handling nullable types (`unwrapNullableType`) has also been refactored and improved to use `typeChecker.getNonNullableType`, which eliminates the manual handling of boolean unions.

We didn't actually have any explicit tests covering enum types, so a new `test/inference/unsupported-types/enum-types.ts` test scenario has been added to explicitly check that we've blocked different forms of enum types. `test/inference/relaxed-types/enum-types-[valid/invalid].ts` covers the behaviour when relaxed types are allowed.

The TS SDK version has also been updated to the latest version available. This will have no impact on the connector.

JIRA: [NLC-22](https://hasurahq.atlassian.net/browse/NLC-22)

### CODE REVIEW

File: ndc-lambda-sdk/src/inference.ts [hunk 1]
```diff
@@ -444,6 +445,31 @@ function deriveSchemaTypeIfScalarType(tsType: ts.Type, context: TypeDerivationCo
...
+  // Handles computed single member enum types: 'enum { OneThing = "test".length }'
+  else if (tsutils.isEnumType(tsType) && tsutils.isSymbolFlagSet(tsType.symbol, ts.SymbolFlags.EnumMember)) {
+    const typeName = context.typeChecker.typeToString(tsType);
+    return deriveRelaxedTypeOrError(typeName, typePath, () => `Enum types are not supported, but one was encountered in ${schema.typePathToString(typePath)} (type: ${typeName})`, allowRelaxedTypes, context);
+  }
+
+  // Note that single member enum types: 'enum { OneThing }' are simplified by the type system
+  // down to literal types (since they can only be a single thing) and are therefore supported via support
+  // for literal types in scalars
+}
```
Comment (sordina) [inline ndc-lambda-sdk/src/inference.ts:471]: Nice

File: ndc-lambda-sdk/test/inference/unsupported-types/unsupported-types.test.ts [hunk 2]
```diff
@@ -271,4 +271,42 @@ describe("unsupported types", function() {
...
+          "Literal union types are not supported, but one was encountered in function 'enumTypes' parameter 'inlineNumberLiteralEnum' (type: 0 | 1 | 2)",
+          "Literal union types are not supported, but one was encountered in function 'enumTypes' parameter 'inlineNumberLiteralEnumMaybe' (type: 0 | 1 | 2)",
+          "Literal union types are not supported, but one was encountered in function 'enumTypes' parameter 'mixedLiteralEnum' (type: MixedLiteralEnum)",
+          "Literal union types are not supported, but one was encountered in function 'enumTypes' parameter 'mixedLiteralEnumMaybe' (type: MixedLiteralEnum)",
+          "Literal union types are not supported, but one was encountered in function 'enumTypes' parameter 'inlineMixedLiteralEnum' (type: true | 1 | \"first\")",
+          "Literal union types are not supported, but one was encountered in function 'enumTypes' parameter 'inlineMixedLiteralEnumMaybe' (type: true | 1 | \"first\")",
+          "Literal union types are not supported, but one was encountered in function 'enumTypes' parameter 'constObjEnum' (type: ConstObjEnum)",
```
Comment (sordina) [inline ndc-lambda-sdk/test/inference/unsupported-types/unsupported-types.test.ts:300]: Great!

**Review by sordina — APPROVED** [review]
Great improvement. Covers a lot of edge cases.
